src/primitives/transaction.cpp

```diff
@@ -62,9 +62,14 @@ uint256 CMutableTransaction::GetHash() const
     return SerializeHash(*this, SER_GETHASH, SERIALIZE_TRANSACTION_NO_WITNESS);
 }
 
-void CTransaction::UpdateHash() const
+void CTransaction::SetHash(const uint256 &tx_hash)
 {
-    *const_cast<uint256*>(&hash) = SerializeHash(*this, SER_GETHASH, SERIALIZE_TRANSACTION_NO_WITNESS);
+    *const_cast<uint256*>(&hash) = tx_hash;
```

Comment:
I don't know if @sipa's suggestion of making a new `CTransaction` constructor that accepts a hash would sufficiently address your concern. (It would not eliminate the risk that some code could make a seemingly harmless `UpdateHash()` call on the wallet transaction and overwrite the correct hash).

Personally, the way I would go about this would be to change `CWalletTx` not to inherit from `CMerkleTx`, and instead just contain either a `variant<CMerkleTx, CCoins>` or an  `optional<CMerkleTx>` member. I can try this in a separate PR (since would create a lot of unrelated code churn in the wallet) if you are amenable to the idea.